README.md
Description
===========

Shorewall is a rather comprehensive and easy-to-use abstraction layer on top of
iptables.


Requirements
============

This cookbook currently uses the `yumrepo` module to install the EPEL
repository, and is therefore CentOS-specific.

The library functions anticipate a network topology in which a cluster of
servers have interconnects over a "private" network which is sufficiently
insecure that a firewall is appropriate to control connections from that
subnet. (This particularly applies to services such as memcached which expect
security to handled at a different layer). However, the module is expected to
remain useful in other scenarios as well.


Capabilities
============

Creates pretty Shorewall configuration files intended to be aesthetically
comparable to hand-written ones.

The following is a typical example of output (in this case, for a rules file):

    #
    # Shorewall version 4 - Rules File
    #
    # For information on the settings in this file, type "man shorewall-rules"
    #
    # The manpage is also online at
    # http://www.shorewall.net/manpages/shorewall-rules.html
    #
    ############################################################################################################################
    #ACTION         SOURCE          DEST            PROTO   DEST    SOURCE          ORIGINAL        RATE            USER/   MARK
    #                                                       PORT    PORT(S)         DEST            LIMIT           GROUP
    #SECTION ESTABLISHED
    #SECTION RELATED
    SECTION NEW

    # Allow all from VM host
    ACCEPT          net:10.0.2.2    fw              -       -       -               -               -               -       -

    # Incoming SSH to firewall
    ACCEPT          all             fw              tcp     22      -               -               -               -       -

    # Allow database load-balancer db1.vguest access to repmgr monitor
    ACCEPT          lan:192.168.123.10 \
                                    fw              tcp     5480    -               -               -               -       -

Note how line continuations are added as necessary to keep column alignment in place.


Usage
=====

Node Variables
--------------

*Important:* Many of these are defined at the `override` level rather than the
`default` level. This is done such that `node[:shorewall][:zones] << { ... }`
works as you'd expect.

* `shorewall/default_interface_settings` - Default settings to be used in
  filling out the `interfaces` file. May be overwritten on a per-interface basis.
* `shorewall/enabled` - Boolean (also accepts string versions of true/false);
  whether we actually start the firewall after configuring it.
* `shorewall/private_ranges` - IP address ranges considered eligible as private
  interconnect addresses.
* `shorewall/zone_hosts/ZONE` - if this starts with `search:`, the remainder is
  used as a search expression to identify hosts which should be considered
  members of this zone (when populating the shorewall `hosts` file). Otherwise,
  it can be a CIDR address (as `192.168.0.0/16` or `0.0.0.0/0`) to refer to a
  subnet.
* `shorewall/zone_interfaces/ZONE` - maps from a shorewall zone name to the
  Ethernet interface serving that zone. If multiple zones are mapped to the
  same interface, then that interface will be distinguished via the shorewall
  `hosts` file.
* `shorewall/rules`, `shorewall/policy`, `shorewall/hosts`,
  `shorewall/interfaces` all correspond directly to the relevant upstream
  configuration files.

For more details, see the `attributes/default.rb` file.


Library
-------

Typical usage from another module is expected to look like the following:

    add_shorewall_rules(
      match_nodes=[
        ['roles:monitoring-server', { :name => 'Monitoring Server' }]
      ],
      rules={
        :description => proc { |data| "Allow #{data["match"]["name"]} #{data["node"]["name"]} access to check_mk" },
        :action => :ACCEPT,
        :source => proc { |data| "lan:#{data["local_address"]}" },
        :dest => :fw,
        :proto => :tcp,
        :dest_port => 6556
      }
    )

...in the above case, we're using the `add_shorewall_rules` helper to add an
`ACCEPT` rule for each host which matches either the `tippr_db::haproxy` recipe
or the `monitoring` role (with different comments depending on which role
matched). If a single host matches twice, only a single rule (for each of its
internal IP addresses) is added.

Notably, any of the values in the `rules` hash can be a block, in which case it
is executed with an argument containing both the match metadata passed to the
`match_nodes` argument and the matched node retrieved by the search operation.

Alternately, an explicit rule (or policy) can be added as follows:

    # Give ALL hosts in lan zone access to logstash
    node.override[:shorewall][:rules] << {
      :description => "Access to logstash web server",
      :action => :ACCEPT,
      :source => :lan,
      :dest => :fw,
      :proto => :tcp,
      :dest_port => 9292
    }

Again: Only address matching one of the networks defined in
`shorewall/private_ranges` will be added by the `add_shorewall_rules` helper.

Poem
====

TODO

Limitations
===========

Patches to address any of these items would be gratefully accepted.

* Includes a hardcoded, non-configurable versions of the `shorewall.conf` file.
* Not all of shorewall's configuration is mapped.
* No thought has been given to IPv6 support.


CHANGELOG.md
# CHANGELOG for networking_basic

This file is used to list changes made in each version of networking_basic.

## 0.1.40:

* Initial release of networking_basic

- - - 
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/default_spec.rb
require 'chefspec'

%w{debian rhel}.each do |platform_family|
  describe "The shorewall::default recipe for #{platform_family}" do
    before (:all) {
      @chef_run = ChefSpec::ChefRunner.new
      @chef_run.node.automatic_attrs["platform_family"] = platform_family
      @chef_run.node.set["shorewall"] = {
        "zone_interfaces" => {"lan" => "eth0"}
      }
      @chef_run.converge 'shorewall::default'
    }

    it 'should install package shorewall' do
      @chef_run.should install_package "shorewall"
    end

    %w{/etc/shorewall/hosts
       /etc/shorewall/zones
       /etc/shorewall/interfaces
       /etc/shorewall/policy
       /etc/shorewall/rules
       /etc/shorewall/shorewall.conf
    }.each do |tpl|
      it "should create file #{tpl}" do
        @chef_run.should create_file tpl
      end
    end

    if platform_family == "debian"
      it "should create file /etc/default/shorewall" do
        @chef_run.should create_file "/etc/default/shorewall"
      end
    end

    {
      "/etc/shorewall/zones"          => "fw      firewall        -               -                       -\n\n" \
                                         "lan     ipv4            -               -                       -",
      "/etc/shorewall/interfaces"     => "lan     eth0            -               -",
      "/etc/shorewall/policy"         => "fw              all             ACCEPT          -               -\n\n" \
                                         "lan             fw              REJECT          -               -\n\n" \
                                         "all             all             REJECT          -               -",
      "/etc/shorewall/rules"          => "ACCEPT          all             fw              tcp     22      -               -               -               -       -",
      "/etc/shorewall/shorewall.conf" => "STARTUP_ENABLED=Yes"
    }.each do |file, content|
      it "should create file #{file} with specific content" do
        @chef_run.should create_file_with_content(file, content)
      end
    end

    it "should enable and start service shorewall on boot" do
      @chef_run.should set_service_to_start_on_boot "shorewall"
      @chef_run.should start_service "shorewall"
    end
  end
end


libraries/ip_search.rb
require 'ipaddr'

def private_addresses_for_node(node_def)
  local_addresses = []
  return local_addresses if node_def['network'] == nil # node may have no ohai data yet
  private_ranges = node['shorewall']['private_ranges'].map { |ip_str| IPAddr.new(ip_str) }
  node_def["network"]["interfaces"].each_pair do |ifname, ifdata|
    if ! ifdata["addresses"] ; then next; end
    ifdata["addresses"].keys.each { |ip_str|
      begin
        if private_ranges.any? { |range| range.include?(IPAddr.new(ip_str)) }
          local_addresses << ip_str
        end
      rescue ArgumentError
        nil # not all addresses are IP; ignore exceptions
      end
    }
  end
  return local_addresses
end

def get_private_addresses(search_criteria, mandatory=false)
  # allow access from app servers
  if Chef::Config[:solo]
    nodes = search(:node, search_criteria)
  else
    nodes = search(:node, search_criteria)
  end
  retval = Array.new
  nodes.each do |matching_node|
    break if matching_node == :node
    private_addresses_for_node(matching_node).each do |interface_address|
      next if not interface_address
      retval << [matching_node, interface_address]
    end
  end
  if mandatory and retval.length == 0
    raise "no matches for mandatory search #{search_criteria}"
  end
  return retval
end

require 'set'
def add_shorewall_rules(match_nodes, rules, mandatory=false)
  done_nodes = Set.new
  match_nodes.each do |search_rule, match_data|
    get_private_addresses(search_rule).each do |matched_node, local_address|
      if done_nodes.include? local_address then
        next
      end
      done_nodes.add local_address
      node.set['shorewall']['rules'] << rules.merge(rules) do |k,v,_|
        if v.is_a? Proc then
          v = v.call({
            :local_address => local_address,
            :match => match_data,
            :node => matched_node
          })
        end
        next v
      end
    end
  end
  if mandatory and done_nodes.length == 0
    raise "no matches for mandatory search"
  end
end

attributes/default.rb
default["shorewall"]["private_ranges"] = ['192.168.0.0/16', '172.16.0.0/12', '10.0.0.0/8']

default["shorewall"]["enabled"] = true

default["shorewall"]["zone_interfaces"]["lan"] = "eth0"

# If we have both net and lan on one interface, search our hosts to assign them to a zone
default["shorewall"]["zone_hosts"]["lan"] = "search:hostname:[* TO *] AND chef_environment:#{node.chef_environment}"
default["shorewall"]["zone_hosts"]["net"] = "0.0.0.0/0"

# Interface default
default["shorewall"]["default_interface_settings"]["broadcast"] = "detect"
default["shorewall"]["default_interface_settings"]["options"] = "tcpflags,blacklist,routefilter,nosmurfs,logmartians,dhcp"

# zones ordered from most specific to most general
default["shorewall"]["zones"] = [
    {"zone" => "fw", "type" => "firewall"},
    {"zone" => "lan", "type" => "ipv4"}
]

# turn "Off" to deactivate IP_FORWARDING
default["shorewall"]["ip_forwarding"] = "On"

default["shorewall"]["policy"] = [
    {"source" => "fw", "dest" => "all", "policy" => "ACCEPT"},
    {"source" => "lan", "dest" => "fw", "policy" => "REJECT", "log" => "DEBUG"},
    {"source" => "all", "dest" => "all", "policy" => "REJECT"}
]

override["shorewall"]["interfaces"] = [ ]

override["shorewall"]["hosts"] = [ ]

default["shorewall"]["rules"] = [
    { "description" => "Incoming SSH to firewall", "source" => "all", "dest" => "fw", "proto" => "tcp", "dest_port" => 22, "action" => "ACCEPT" },
    { "description" => "Incoming check_mk to firewall", "source" => "lan", "dest" => "fw", "proto" => "tcp", "dest_port" => 6556, "action" => "ACCEPT" }
]